### PR TITLE
Add support for sub heading in extension views

### LINKED ISSUE #9903: Implement vscode TreeView description
### Feature Description:

I would like to implement TreeView.description from the vscode api. This appears as a small subheading next to the panel title.

https://code.visualstudio.com/api/references/vscode-api#TreeView%3CT%3E

### PR DESCRIPTION
**Pull request by aaronrhodes**
#### What it does

Adds the description field to TreeView from the vscode API.

resolves #9903 

#### How to test

Add a plugin that utilises the TreeView description field.

There is a modified version of the vscode tree-view-sample available in the following repo:

https://github.com/aaronrhodes/vscode-extension-samples

build, and copy or symlink this example into the plugins directory and select a file using the "file explorer" panel in the explorer tab.

Screenshot 2021-08-16 at 11 03 57

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/core/src/browser/shell/description-widget.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/********************************************************************************
+ * Copyright (C) 2017 TypeFox and others.
```
Comment (msujew) [inline packages/core/src/browser/shell/description-widget.ts:2]: As you created this file yourself, please change the copyright notice to you/your company.

File: packages/core/src/browser/shell/description-widget.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+export interface DescriptionWidget {
+    description: string;
+    onDescriptionChange: Emitter<void>;
+}
+
+export namespace DescriptionWidget {
+    /* eslint-disable @typescript-eslint/no-explicit-any */
+    export function is(arg: any): arg is DescriptionWidget {
+        return arg !== undefined && !!arg['onDescriptionChange'];
+    }
```
Comment (msujew) [inline packages/core/src/browser/shell/description-widget.ts:28]: Suggestion: We generally try not to use `any`, even in `is` methods. Change this to `Object | undefined` or `unknown`.

**Review by msujew — COMMENTED** [review]
Thank you for your contribution, the changes look good to me so far. The description property does its job. 

FYI the title of your PR is a bit misleading, as you're adding support for the description property for views contributed by extensions, not only for outlines.

I have a few remarks before approving.

File: packages/core/src/browser/shell/description-widget.ts (hunk 2 above)
Comment (vince-fugnitto) [inline packages/core/src/browser/shell/description-widget.ts:28]: I wasn't so sure the file was necessary, and instead it could be directly added to the `tree-view`.
Adding it under `shell/` is also a bit awkward.

File: packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts [hunk 3]
```diff
@@ -112,6 +116,17 @@ export class PluginViewWidget extends Panel implements StatefulWidget {
...
     }
 
+    private _description: string = '';
+    get description(): string {
+        return this._description;
+    }
+
+    set description(description: string) {
+        this._description = description;
+        this.onDescriptionChange.fire();
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:126]: Should be moved under `this.updateWidgetDescription()`.

File: packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts [hunk 4]
```diff
@@ -121,14 +136,25 @@ export class PluginViewWidget extends Panel implements StatefulWidget {
...
     }
 
+    private updateWidgetDescription(): void {
+        const widget = this.widgets[0];
+        if (widget) {
+            if (DescriptionWidget.is(widget)) {
+                widget.description = this._description;
+            }
+        }
+    }
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:146]:
```suggestion
    private updateWidgetDescription(): void {
        const widget = this.widgets[0];
        if (widget) {
            if (DescriptionWidget.is(widget)) {
                widget.description = this._description;
                widget.onDescriptionChange.fire();
            }
        }
    }
```

File: packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts [hunk 5]
```diff
@@ -50,6 +52,8 @@ export class PluginViewWidget extends Panel implements StatefulWidget {
         this.node.style.height = '100%';
     }
 
+    public onDescriptionChange: Emitter<void> = new Emitter<void>();
```
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:55]: Can be removed, no need to duplicate the event and instead use the widget directly.

**Review by vince-fugnitto — COMMENTED** [review]
I noticed a couple of things:
- the commit message needs to be updated, we mention `outline-view` but I believe we mean all view-container-parts.
- the description is only visible if the view-container-part is currently expanded in vscode - I believe we should align.

File: packages/core/src/browser/shell/description-widget.ts (hunk 2 above)
Comment (aaronrhodes) [reply packages/core/src/browser/shell/description-widget.ts:28]: I copied this implementation from a few examples in the rest of the code and did initially try and use `unknown`, but the compiler wouldn't let me use the string object accessor. Will update to `Object | undefined`.

Comment (msujew) [reply packages/core/src/browser/shell/description-widget.ts:28]: For `unknown` you'll have to do this:

```ts
return typeof arg === 'object' && arg && !!arg['onDescriptionChange'];
```

Comment (aaronrhodes) [reply packages/core/src/browser/shell/description-widget.ts:28]: I still get compiler errors about any having an implicit any type whichever way I try this..

Comment (aaronrhodes) [reply packages/core/src/browser/shell/description-widget.ts]: I'm now moved this file into `view-container.ts` as it only really applies to the container widgets

File: packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts (hunk 3 above)
Comment (aaronrhodes) [reply packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:126]: Have now removed this block of code as it was redundant - I use the `setWidgetMessage` flow as an example when first writing this but it's now obvious that description is more of an upward flowing property and has no utility or path to be passed down.

File: packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts (hunk 5 above)
Comment (aaronrhodes) [reply packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:55]: As mentioned above, the wrapped widget actually doesn't have the `DescriptionWidget` interface implemented, but `plugin-view-widget` does so is needed

**Comment by aaronrhodes** [comment]
> I noticed a couple of things:
> 
> * the commit message needs to be updated, we mention `outline-view` but I believe we mean all view-container-parts.
> * the description is only visible if the view-container-part is currently expanded in vscode - I believe we should align.

thanks for the feedback! Have updated the commit message and the subheading now hides and shows when the panel is collapsed/expanded respectively.

File: packages/core/src/browser/shell/description-widget.ts (hunk 2 above)
Comment (vince-fugnitto) [reply packages/core/src/browser/shell/description-widget.ts]: @aaronrhodes you can do the following to avoid using `any`:

```ts
export namespace DescriptionWidget {
    export function is(arg: Object | undefined): arg is DescriptionWidget {
        return !!arg && typeof arg === 'object' && 'onDidDescriptionChange' in arg;
    }
}
```

Comment (aaronrhodes) [reply packages/core/src/browser/shell/description-widget.ts]: 👍

File: packages/core/src/browser/view-container.ts [hunk 6]
```diff
@@ -853,15 +873,33 @@ export class ViewContainerPart extends BaseWidget {
...
+        subHeading.classList.add('subheading');
+
         const updateTitle = () => title.innerText = this.wrapped.title.label;
         const updateCaption = () => title.title = this.wrapped.title.caption || this.wrapped.title.label;
+        const updateSubHeading = () => {
+            if (DescriptionWidget.is(this.wrapped) && !this.collapsed) {
+                subHeading.innerText = !!this.wrapped.description ? this.wrapped.description : '';
+            } else {
+                subHeading.innerText = '';
+            }
```
Comment (msujew) [inline packages/core/src/browser/view-container.ts:887]: Nitpick: We can simplify this

```suggestion
            subHeading.innerText = DescriptionWidget.is(this.wrapped) && !this.collapsed && this.wrapped.description || '';
```

File: packages/core/src/browser/view-container.ts [hunk 7]
```diff
@@ -699,6 +712,12 @@ export class ViewContainerPart extends BaseWidget {
         this.wrapped.title.changed.connect(fireTitleChanged);
         this.toDispose.push(Disposable.create(() => this.wrapped.title.changed.disconnect(fireTitleChanged)));
 
+        if (DescriptionWidget.is(this.wrapped)) {
+            const fireSubHeadingChanged = () => this.onSubHeadingChangedEmitter.fire(undefined);
+            const sub = (this.wrapped as DescriptionWidget)?.onDescriptionChange.event(fireSubHeadingChanged);
```
Comment (vince-fugnitto) [inline packages/core/src/browser/view-container.ts:717]: Update to a more meaningful name than `sub`?

File: packages/core/src/browser/style/view-container.css [hunk 8]
```diff
@@ -75,10 +75,19 @@
...
+    white-space: nowrap;
+    text-overflow: ellipsis;
+}
+
+.theia-view-container .part > .header .subheading {
     flex: 1;
     overflow: hidden;
     white-space: nowrap;
     text-overflow: ellipsis;
+    padding-left: 12px;
```
Comment (vince-fugnitto) [inline packages/core/src/browser/style/view-container.css:88]: We should compute this value I believe rather than hardcoding.
Example:

https://github.com/eclipse-theia/theia/blob/a6422c39d6c09095f2bdc8cb4af4427e6bc78082/packages/debug/src/browser/style/index.css#L142

File: packages/core/src/browser/view-container.ts (hunk 6 above)
Comment (aaronrhodes) [reply packages/core/src/browser/view-container.ts:887]: 👍

File: packages/core/src/browser/view-container.ts [hunk 9]
```diff
@@ -673,6 +684,8 @@ export class ViewContainerPart extends BaseWidget {
     readonly onTitleChanged = this.onTitleChangedEmitter.event;
     protected readonly onDidFocusEmitter = new Emitter<this>();
     readonly onDidFocus = this.onDidFocusEmitter.event;
+    protected readonly onSubHeadingChangedEmitter = new Emitter<void>();
+    readonly onSubHeadingChanged = this.onSubHeadingChangedEmitter.event;
```
[the remark this answers is not in the document]
Comment (aaronrhodes) [reply packages/core/src/browser/view-container.ts:688]: 👍

File: packages/core/src/browser/style/view-container.css (hunk 8 above)
Comment (aaronrhodes) [reply packages/core/src/browser/style/view-container.css:88]: Great suggestion, thanks!

File: packages/core/src/browser/view-container.ts [hunk 10]
```diff
@@ -673,6 +684,8 @@ export class ViewContainerPart extends BaseWidget {
     readonly onTitleChanged = this.onTitleChangedEmitter.event;
     protected readonly onDidFocusEmitter = new Emitter<this>();
     readonly onDidFocus = this.onDidFocusEmitter.event;
+    protected readonly onDidSubHeadingChangedEmitter = new Emitter<void>();
+    readonly onDidSubHeadingChanged = this.onDidSubHeadingChangedEmitter.event;
```
Comment (colin-grant-work) [inline packages/core/src/browser/view-container.ts:688]: I think `onDidChangeSubHeading(Emitter)` would conform more closely to the guidelines on [event naming](https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#event_names).

File: packages/core/src/browser/view-container.ts [hunk 11]
```diff
@@ -699,6 +712,11 @@ export class ViewContainerPart extends BaseWidget {
         this.wrapped.title.changed.connect(fireTitleChanged);
         this.toDispose.push(Disposable.create(() => this.wrapped.title.changed.disconnect(fireTitleChanged)));
 
+        if (DescriptionWidget.is(this.wrapped)) {
+            const fireSubHeadingChanged = () => this.onDidSubHeadingChangedEmitter.fire(undefined);
+            this.toDispose.push((this.wrapped as DescriptionWidget)?.onDidDescriptionChange.event(fireSubHeadingChanged));
```
Comment (colin-grant-work) [inline packages/core/src/browser/view-container.ts:717]: Since you've performed a type check already, you shouldn't need the `as` cast or the `?.` conditional accessor.

Why the change in terminology here from `onDidChangeDescription` to `onDidChangeSubHeading`? I think I'd prefer that we use the `description`-type names throughout and leave the fact that they're rendered as a subheading as an implementation detail.

File: packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts [hunk 12]
```diff
@@ -50,6 +52,8 @@ export class PluginViewWidget extends Panel implements StatefulWidget {
         this.node.style.height = '100%';
     }
 
+    public onDidDescriptionChange: Emitter<void> = new Emitter<void>();
```
Comment (colin-grant-work) [inline packages/plugin-ext/src/main/browser/view/plugin-view-widget.ts:55]:
```suggestion
    onDidChangeDescription: Emitter<void> = new Emitter<void>();
```
